File: assignment.js
// Define classes implementing interfaces
var Keyboard = /** @class */ (function () {
    function Keyboard() {
    }
    Keyboard.prototype.input = function () {
        return "Keyboard input";
    };
    return Keyboard;
}());
var Mouse = /** @class */ (function () {
    function Mouse() {
    }
    Mouse.prototype.input = function () {
        return "Mouse input";
    };
    return Mouse;
}());
var SSD = /** @class */ (function () {
    function SSD() {
    }
    SSD.prototype.store = function () {
        return "SSD storing data";
    };
    SSD.prototype.retrieve = function () {
        return "SSD retrieving data";
    };
    return SSD;
}());
var HDD = /** @class */ (function () {
    function HDD() {
    }
    HDD.prototype.store = function () {
        return "HDD storing data";
    };
    HDD.prototype.retrieve = function () {
        return "HDD retrieving data";
    };
    return HDD;
}());
var CPU = /** @class */ (function () {
    function CPU() {
    }
    CPU.prototype.process = function () {
        return "CPU processing data";
    };
    return CPU;
}());
var GPU = /** @class */ (function () {
    function GPU() {
    }
    GPU.prototype.process = function () {
        return "GPU processing data";
    };
    return GPU;
}());
var Monitor = /** @class */ (function () {
    function Monitor() {
    }
    Monitor.prototype.output = function () {
        return "Monitor output";
    };
    return Monitor;
}());
var Printer = /** @class */ (function () {
    function Printer() {
    }
    Printer.prototype.output = function () {
        return "Printer output";
    };
    return Printer;
}());
// WiFi & Bluetooth classes
var WiFi = /** @class */ (function () {
    function WiFi() {
    }
    WiFi.prototype.connect = function () {
        return "WiFi connected";
    };
    WiFi.prototype.disconnect = function () {
        return "WiFi disconnected";
    };
    return WiFi;
}());
var Bluetooth = /** @class */ (function () {
    function Bluetooth() {
    }
    Bluetooth.prototype.connect = function () {
        return "Bluetooth connected";
    };
    Bluetooth.prototype.disconnect = function () {
        return "Bluetooth disconnected";
    };
    return Bluetooth;
}());
// Define Computer class
var Computer = /** @class */ (function () {
    function Computer(inputDevice, storageDevice, processingDevice, outputDevice, wifi, bluetooth) {
        this._inputDevice = inputDevice;
        this._storageDevice = storageDevice;
        this._processingDevice = processingDevice;
        this._outputDevice = outputDevice;
        this._wifi = wifi;
        this._bluetooth = bluetooth;
    }
    Computer.prototype.setInputDevice = function (device) {
        this._inputDevice = device;
    };
    Computer.prototype.setStorageDevice = function (device) {
        this._storageDevice = device;
    };
    Computer.prototype.setProcessingDevice = function (device) {
        this._processingDevice = device;
    };
    Computer.prototype.setOutputDevice = function (device) {
        this._outputDevice = device;
    };
    Computer.prototype.setWiFi = function (wifi) {
        this._wifi = wifi;
    };
    Computer.prototype.setBluetooth = function (bt) {
        this._bluetooth = bt;
    };
    Computer.prototype.performInput = function () {
        return this._inputDevice.input();
    };
    Computer.prototype.performStorage = function () {
        return this._storageDevice.store() + " and " + this._storageDevice.retrieve();
    };
    Computer.prototype.performProcessing = function () {
        return this._processingDevice.process();
    };
    Computer.prototype.performOutput = function () {
        return this._outputDevice.output();
    };
    Computer.prototype.connectWiFi = function () {
        if (this._wifi) {
            return this._wifi.connect();
        }
        return "No WiFi support";
    };
    Computer.prototype.connectBluetooth = function () {
        if (this._bluetooth) {
            return this._bluetooth.connect();
        }
        return "No Bluetooth support";
    };
    return Computer;
}());
// Example usage
var keyboard = new Keyboard();
var mouse = new Mouse();
var ssd = new SSD();
var cpu = new CPU();
var monitor = new Monitor();
var wifi = new WiFi();
var bluetooth = new Bluetooth();
var computer = new Computer(keyboard, ssd, cpu, monitor, wifi, bluetooth);
console.log(computer.performInput()); // Outputs: Keyboard input
console.log(computer.performStorage()); // Outputs: SSD storing data and SSD retrieving data
console.log(computer.performProcessing()); // Outputs: CPU processing data
console.log(computer.performOutput()); // Outputs: Monitor output
console.log(computer.connectWiFi()); // Outputs: WiFi connected
console.log(computer.connectBluetooth()); // Outputs: Bluetooth connected
// Change the input device
computer.setInputDevice(mouse);
console.log(computer.performInput()); // Outputs: Mouse input
// Create a computer without WiFi and Bluetooth support
var basicComputer = new Computer(keyboard, ssd, cpu, monitor);
console.log(basicComputer.connectWiFi()); // Outputs: No WiFi support
console.log(basicComputer.connectBluetooth()); // Outputs: No Bluetooth support

File: python/assignment.py
from abc import ABC, abstractmethod

# Define interfaces
class InputDevice(ABC):
    @abstractmethod
    def input(self) -> str:
        pass

class StorageDevice(ABC):
    @abstractmethod
    def store(self) -> str:
        pass

    @abstractmethod
    def retrieve(self) -> str:
        pass

class ProcessingDevice(ABC):
    @abstractmethod
    def process(self) -> str:
        pass

class OutputDevice(ABC):
    @abstractmethod
    def output(self) -> str:
        pass
class WiFi(ABC):
    def connectToWifi(self) -> str:
        pass
    
    def disconnectFromWifi(self) -> str:
        pass
class Bluetooth(ABC):
    def connectToBluetooth(self) -> str:
        pass
    
    def disconnectFromBluetooth(self) -> str:
        pass    

class Keyboard(InputDevice):
    def input(self) -> str:
        return "Keyboard input"

class Mouse(InputDevice):
    def input(self) -> str:
        return "Mouse input"

class SSD(StorageDevice):
    def store(self) -> str:
        return "SSD storing data"
    
    def retrieve(self) -> str:
        return "SSD retrieving data"

class HDD(StorageDevice):
    def store(self) -> str:
        return "HDD storing data"
    
    def retrieve(self) -> str:
        return "HDD retrieving data"

class CPU(ProcessingDevice):
    def process(self) -> str:
        return "CPU processing data"

class GPU(ProcessingDevice):
    def process(self) -> str:
        return "GPU processing data"

class Monitor(OutputDevice):
    def output(self) -> str:
        return "Monitor output"

class Printer(OutputDevice):
    def output(self) -> str:
        return "Printer output"



# Define Computer class
class Computer:
    def __init__(
        self,
        input_device: InputDevice,
        storage_device: StorageDevice,
        processing_device: ProcessingDevice,
        output_device: OutputDevice,
       
    ):
        self._input_device = input_device
        self._storage_device = storage_device
        self._processing_device = processing_device
        self._output_device = output_device
       
    def set_input_device(self, device: InputDevice) -> None:
        self._input_device = device

    def set_storage_device(self, device: StorageDevice) -> None:
        self._storage_device = device

    def set_processing_device(self, device: ProcessingDevice) -> None:
        self._processing_device = device

    def set_output_device(self, device: OutputDevice) -> None:
        self._output_device = device

    
    def perform_input(self) -> str:
        return self._input_device.input()

    def perform_storage(self) -> str:
        return f"{self._storage_device.store()} and {self._storage_device.retrieve()}"

    def perform_processing(self) -> str:
        return self._processing_device.process()

    def perform_output(self) -> str:
        return self._output_device.output()

    def connect_wifi(self) -> str:
        if self._wifi:
            return self._wifi.connect()
        return "No WiFi support"

    def connect_bluetooth(self) -> str:
        if self._bluetooth:
            return self._bluetooth.connect()
        return "No Bluetooth support"
    



class Laptop(Computer):
    def __init__(
        self,
        wifi: WiFi, 
        bluetooth: Bluetooth ,
       
    ):
        self.wifi = wifi
        self.bluetooth = bluetooth
        
    def connectToBluetooth(self) -> str:
        return "Bluetooth connected"
    
    def disconnectFromBluetooth(self) -> str:
        return "Bluetooth disconnected"
    def connectToWifi(self) -> str:
        return "WiFi connected"
    
    def disconnectFromWifi(self) -> str:
        return "WiFi disconnected"





# Example usage
keyboard = Keyboard()
mouse = Mouse()
ssd = SSD()
cpu = CPU()
monitor = Monitor()
wifi = WiFi()
bluetooth = Bluetooth()

computer = Computer(keyboard, ssd, cpu, monitor)
print(computer.perform_input())      # Outputs: Keyboard input
print(computer.perform_storage())    # Outputs: SSD storing data and SSD retrieving data
print(computer.perform_processing()) # Outputs: CPU processing data
print(computer.perform_output())     # Outputs: Monitor output
# Change the input device
computer.set_input_device(mouse)
print(computer.perform_input())      # Outputs: Mouse input
# Create a computer without WiFi and Bluetooth support
basic_computer = Computer(keyboard, ssd, cpu, monitor)



Laptop1=Laptop(wifi, bluetooth)
print(Laptop1.connectToBluetooth())  # Outputs: Bluetooth connected
print(Laptop1.disconnectFromBluetooth())  # Outputs: Bluetooth disconnected
print(Laptop1.connectToWifi())  # Outputs: WiFi connected
print(Laptop1.disconnectFromWifi()) # Outputs:WiFi disconnected
from abc import ABC, abstractmethod

# Define interfaces
class InputDevice(ABC):
    @abstractmethod
    def input(self) -> str:
        pass

class StorageDevice(ABC):
    @abstractmethod
    def store(self) -> str:
        pass

    @abstractmethod
    def retrieve(self) -> str:
        pass

class ProcessingDevice(ABC):
    @abstractmethod
    def process(self) -> str:
        pass

class OutputDevice(ABC):
    @abstractmethod
    def output(self) -> str:
        pass
class WiFi(ABC):
    def connectToWifi(self) -> str:
        pass
    
    def disconnectFromWifi(self) -> str:
        pass
class Bluetooth(ABC):
    def connectToBluetooth(self) -> str:
        pass
    
    def disconnectFromBluetooth(self) -> str:
        pass    

class Keyboard(InputDevice):
    def input(self) -> str:
        return "Keyboard input"

class Mouse(InputDevice):
    def input(self) -> str:
        return "Mouse input"

class SSD(StorageDevice):
    def store(self) -> str:
        return "SSD storing data"
    
    def retrieve(self) -> str:
        return "SSD retrieving data"

class HDD(StorageDevice):
    def store(self) -> str:
        return "HDD storing data"
    
    def retrieve(self) -> str:
        return "HDD retrieving data"

class CPU(ProcessingDevice):
    def process(self) -> str:
        return "CPU processing data"

class GPU(ProcessingDevice):
    def process(self) -> str:
        return "GPU processing data"

class Monitor(OutputDevice):
    def output(self) -> str:
        return "Monitor output"

class Printer(OutputDevice):
    def output(self) -> str:
        return "Printer output"



# Define Computer class
class Computer:
    def __init__(
        self,
        input_device: InputDevice,
        storage_device: StorageDevice,
        processing_device: ProcessingDevice,
        output_device: OutputDevice,
       
    ):
        self._input_device = input_device
        self._storage_device = storage_device
        self._processing_device = processing_device
        self._output_device = output_device
       
    def set_input_device(self, device: InputDevice) -> None:
        self._input_device = device

    def set_storage_device(self, device: StorageDevice) -> None:
        self._storage_device = device

    def set_processing_device(self, device: ProcessingDevice) -> None:
        self._processing_device = device

    def set_output_device(self, device: OutputDevice) -> None:
        self._output_device = device

    
    def perform_input(self) -> str:
        return self._input_device.input()

    def perform_storage(self) -> str:
        return f"{self._storage_device.store()} and {self._storage_device.retrieve()}"

    def perform_processing(self) -> str:
        return self._processing_device.process()

    def perform_output(self) -> str:
        return self._output_device.output()

    def connect_wifi(self) -> str:
        if self._wifi:
            return self._wifi.connect()
        return "No WiFi support"

    def connect_bluetooth(self) -> str:
        if self._bluetooth:
            return self._bluetooth.connect()
        return "No Bluetooth support"
    



class Laptop(Computer):
    def __init__(
        self,
        wifi: WiFi, 
        bluetooth: Bluetooth ,
       
    ):
        self.wifi = wifi
        self.bluetooth = bluetooth
        
    def connectToBluetooth(self) -> str:
        return "Bluetooth connected"
    
    def disconnectFromBluetooth(self) -> str:
        return "Bluetooth disconnected"
    def connectToWifi(self) -> str:
        return "WiFi connected"
    
    def disconnectFromWifi(self) -> str:
        return "WiFi disconnected"





# Example usage
keyboard = Keyboard()
mouse = Mouse()
ssd = SSD()
cpu = CPU()
monitor = Monitor()
wifi = WiFi()
bluetooth = Bluetooth()

computer = Computer(keyboard, ssd, cpu, monitor)
print(computer.perform_input())      # Outputs: Keyboard input
print(computer.perform_storage())    # Outputs: SSD storing data and SSD retrieving data
print(computer.perform_processing()) # Outputs: CPU processing data
print(computer.perform_output())     # Outputs: Monitor output
# Change the input device
computer.set_input_device(mouse)
print(computer.perform_input())      # Outputs: Mouse input
# Create a computer without WiFi and Bluetooth support
basic_computer = Computer(keyboard, ssd, cpu, monitor)



Laptop1=Laptop(wifi, bluetooth)
print(Laptop1.connectToBluetooth())  # Outputs: Bluetooth connected
print(Laptop1.disconnectFromBluetooth())  # Outputs: Bluetooth disconnected
print(Laptop1.connectToWifi())  # Outputs: WiFi connected
print(Laptop1.disconnectFromWifi()) # Outputs:WiFi disconnected